pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/dictionary/BaseOffHeapMutableDictionary.java

```diff
@@ -0,0 +1,411 @@
+ * AtomicReference. This ensures that readers always get either the new version of these objects or the old version,
+ * but not some inconsistent versions of these.
+ *
+ * It should be noted that this class assumes that there is one writer and multiple readers of the dictionary. It is
+ * NOT safe for a multiple writer scenario.
+ *
+ * TODO
+ * - It may be useful to implement a way to stop adding new items when the the number of buffers reaches a certain
+ *   threshold. In this case, we could close the realtime segment, and start a new one with bigger buffers.
+ */
+public abstract class BaseOffHeapMutableDictionary extends MutableDictionary {
+  // Percent of cardinality that we expect to hold in the overflow hash table.
+  private static final int OVERFLOW_THRESHOLD_PERCENT = 1;
+
+  // expansionMultiple setting as we add new buffers. A setting of 1 sets the new buffer to be
+  // the same size as the last one added. Setting of 2 allocates a buffer twice as big as the
+  // previous one. It is assumed that these values are powers of 2. It is a good idea to restrict
+  // these to 1 or 2, but not higher values. This array can be arbitrarily long. If the number of
+  // buffers exceeds the number of elements in the array, we use the last value.
+  private static final int[] EXPANSION_MULTIPLES = new int[] {1, 1, 2, 2, 2};
+
+  // Number of columns in each row of an IntBuffer.
+  private static final int NUM_COLUMNS = 3;
+
+  // Whether to start with 0 off-heap storage. Items are added to the overflow map first, until it reaches
+  // a threshold, and then the off-heap structures are allocated.
+  private static final boolean HEAP_FIRST = true;
+
+  private final int _maxItemsInOverflowHash;
+
+  // Number of entries in the dictionary. Max dictId is _numEntries-1.
+  private int _numEntries;
+
+  // We keep a list of PinotDataBuffer items from which we get the IntBuffer items, so
+  // that we can call close() on these.
+  private List<PinotDataBuffer> _pinotDataBuffers = new ArrayList<>();
+  private final int _sizeOfFirstBuf;
+
+  /**
+   * A class to hold all the objects needed for the reverse mapping.
+   */
+  private static class ValueToDictId {
+    // Each iBuf layout is as described in comments above.
+    private final List<IntBuffer> _iBufList;
+    // The map should be a concurrent one.
+    private final Map<Object, Integer> _overflowMap;
+
+    private ValueToDictId(List<IntBuffer> iBufList, Map<Object, Integer> overflowMap) {
+      _iBufList = iBufList;
+      _overflowMap = overflowMap;
+    }
+
+    private List<IntBuffer> getIBufList() {
+      return _iBufList;
+    }
+    private Map<Object, Integer> getOverflowMap() {
+      return _overflowMap;
+    }
+  }
+
+  private AtomicReference<ValueToDictId> _valueToDict = new AtomicReference<>();
+
+  protected BaseOffHeapMutableDictionary(int estimatedCardinality) {
```

Comment:
Like I mentioned before, the current plan is to start with a guess and then learn from the previous segment. Mechanisms TBD.